### PR TITLE
Added threshold, otsu level

### PR DESCRIPTION
**Pull request by akonneker**
I spent some wasted time chasing a nonexistent bug trying to get the results to match matlab's... but it should work now. Haven't used it on actual images yet, but I will "soon".

The style may be terrible, but I am a rank amateur when it comes to rust, so I hope you'll forgive me.

I'll be swamped for a little while yet, but when I can carve out time I'm going to port some matlab code for more interesting algorithms.

I'll also add some other thresholding options eventually, too.

### CODE REVIEW

File: src/contrast.rs [hunk 1]
```diff
@@ -7,9 +8,86 @@ use image::{
     Luma
 };
 
-/// Returns the cumulative histogram of grayscale values in an 8bpp
+/// Returns the Otsu threshold level of an 8bpp image.
+/// This threshold will optimally binarize an image that
+/// meets certain criteria.
+pub fn otsu_level<I>(image: &I) -> u8
```
Comment (theotherphil) [inline src/contrast.rs]: "meets certain criteria". Which criteria?

File: src/contrast.rs [hunk 2]
```diff
@@ -7,9 +8,86 @@ use image::{
...
+    for i in 1..levels {
+        pixel_count += hist[i] as f64;
+        histc[i] = histc[i-1] + hist[i];
+        meanc[i] = meanc[i-1] + (hist[i] as f64)*(i as f64);
+    }
+
+    let mut sigma_max:f64 = -1.0;
+    let mut otsu:f64 = 0.0;
+    let mut otsu2:f64 = 0.0;
+
```
Comment (theotherphil) [inline src/contrast.rs]: Minor nitpick: most rust uses a space after the colons here. Also, you could just write let mut x = 0f64.

File: src/contrast.rs [hunk 3]
```diff
@@ -7,9 +8,86 @@ use image::{
...
+    }
+    ( (otsu + otsu2) / 2.0 ).ceil() as u8
+}
+
+/// Returns a binarized image from an input 8bpp grayscale image
+/// obtained by applying the given image
+pub fn threshold<I>(image: &I, thresh: u8) -> GrayImage
+    where I: GenericImage<Pixel=Luma<u8>> {
+    let mut out: GrayImage = ImageBuffer::new(image.width(), image.height());
+     for y in 0..image.height() {
```
Comment (theotherphil) [inline src/contrast.rs]: I suspect you can get away without the type annotation here.

File: src/contrast.rs [hunk 4]
```diff
@@ -7,9 +8,86 @@ use image::{
...
+            for x in 0..image.width() {
+                if image.get_pixel(x, y)[0] as u8 <= thresh {
+                    out.put_pixel(x, y, Luma([0 as u8]));
+                } else {
```
Comment (theotherphil) [inline src/contrast.rs]: And here.

File: src/contrast.rs [hunk 5]
```diff
@@ -7,9 +8,86 @@ use image::{
...
+                    out.put_pixel(x, y, Luma([255 as u8]));
+                }
```
Comment (theotherphil) [inline src/contrast.rs]: None of these "as u8" are necessary.

File: src/contrast.rs [hunk 6]
```diff
@@ -7,9 +8,86 @@ use image::{
...
+            }
+        }
+    out
+}
+
+/// Mutates given image to form a binarized version produced by applying
+/// the given threshold
+pub fn threshold_mut<I>(image: &mut I, thresh: u8)
+    where I: GenericImage<Pixel=Luma<u8>>  {
+     for y in 0..image.height() {
```
Comment (theotherphil) [inline src/contrast.rs]: When there's a pair of f and f_mut, I'd prefer that f_mut contain the full logic, and the body of f just be: "let out = ImageBuffer::new(...); out.copy_from(input, 0, 0); f(out); out". See e.g equalize_histogram and equalize_histogram_mut.

File: src/contrast.rs [hunk 7]
```diff
@@ -83,6 +175,51 @@ mod test {
...
+                255u8,  255u8,  255u8,  255u8]).unwrap();
+
+        let test = threshold(&before, 125u8);
+
+        for y in 0..after.height() {
+            for x in 0..after.width() {
+                let check = after.get_pixel(x, y)[0] as usize;
+                let test_pixel = test.get_pixel(x, y)[0] as usize;
+                assert_eq!(test_pixel, check);
+            }
```
Comment (theotherphil) [inline src/contrast.rs]: assert_pixels_eq will do this more succinctly: assert_pixels_eq!(test, after). I'd use "actual" for "test" and "expected" for "after" here, as it's slightly clearer what each value is for.

**Comment by theotherphil** [comment]
Looks good, thanks! I've added a selection of pretty pedantic comments - I hope you don't mind the nitpicking. Nothing major, so I'll merge this either when you make the changes or in a couple of days if you don't have the time/inclination to make them.

**Comment by theotherphil** [comment]
Any other stuff you fancy adding would be great. If the other algorithms you're considering porting are "interesting" in non-image-processing ways then maybe we should put them in a currently nonexistent computer vision sister project.

**Comment by akonneker** [comment]
I'll happily make the changes suggested. As I said, my rust experience is essentially nil, so I appreciate any suggestions. 

What I'm going to implement first are some things I'm already working with/on. One is a [phase congruency filter] (http://cvpr11.cecs.anu.edu.au/files/PhaseCongCVPR.pdf). This would be a nice addition to the library, I think.

The other is an RBG-D SLAM algorithm like [kinectfusion](https://github.com/Nerei/kinfu_remake) or [dvo-slam](https://github.com/tum-vision/dvo_slam).

So SLAM would definitely need a new home. I'll start on that in earnest pretty soon, but it'll be a few weeks at the earliest before it's finished. So I'll let you know.

**Comment by theotherphil** [comment]
I'd not heard of phase congruency filters before - looks interesting. I'm also planning on implementing a few vision algorithms soon (http://www.edwardrosten.com/work/taylor_2009_robust.pdf, https://www.cs.cmu.edu/~efros/courses/LBMV07/Papers/viola-cvpr-01.pdf) but it doesn't look like there'll be much overlap in required code yet.

File: src/contrast.rs (hunk 3 above)
Comment (akonneker) [inline src/contrast.rs]: These annotations appear to be necessary to avoid type mismatch errors.

**Comment by theotherphil** [comment]
Great, thanks!

**Pull request merged**